### PR TITLE
feat(cli,config): info command prints stored/read cluster metadata

### LINKED ISSUE #454: CLI: add cluster-info command
## Current Behavior / Proposed Behavior

The CLI version flag only prints the CLI installer version. 

```bash
$ opstrace --version 
   Version vX.Y.Z (CLI installer version)
```

When upgrading the Opstracer controller with the current work-in-progress `upgrade` command, the installer and controller versions will diverge. We need a way to show this information to the user.

```bash
$ opstrace cluster-info CLOUD_PROVIDER CLUSTER_NAME
   cluster/infra: version A.B.C
   controller: version X.Y.Z
   installer: version A.B.C
```

## Context

Cluster or infra version refers to all the components not created by the opstrace-controller, for example, the cloud instances, the VPC, and the Kubernetes cluster.

The installer version refers to the Opstrace CLI the user downloads and executes to run commands.

Controller version refers to the Docker image tag used in the Opstrace controller deployment installed in the Kubernetes cluster.

## Possible Technical Solution

Add cluster/infra (CLI version that created the cluster infrastructure) and controller version in the controller config map. Add version subcommand to print cluster/infra version, controller version, and installer CLI version.

### PR DESCRIPTION
**Pull request by estroz**
`opstrace info <aws|gcp> <cluster name>` will print Opstrace and Kubernetes cluster info in the following format:

```
cluster info:
  cluster_version: <kubernetes version>
  controller_version: <image tag>
  installer_version: <installer commit string>
```

Installer version is stored in the controller's configmap and updated on each upgrade event, while other version information is read from k8s cluster configuration at runtime.

Closes #454

Might also address part of #1023

/cc @sreis 

# Have you...

* [x] Discussed your change with a project contributor in an issue yet?
* [x] Added an explanation of what your changes do?
* [x] Written new tests for your changes?
* [x] Thought about which docs need updating?

### CODE REVIEW

**Comment by estroz** [comment]
will update CLI docs once the command name/output is agreed on.

File: packages/cli/src/info.ts
Comment (estroz) [inline packages/cli/src/info.ts:53]: These kubeconfig API calls work, but is there a better library to use, like update reducers?

File: packages/cli/src/index.ts
Comment (jgehrcke) [inline packages/cli/src/index.ts:123]: :+1: thanks

File: packages/cli/src/index.ts
Comment (jgehrcke) [inline packages/cli/src/index.ts:184]: I like that. 

Please replace "cluster" with "instance", for terminology consistency (we made that decision a while ago :shrug:).

Let's maybe say "version information" because that's all there is for now.

**Comment by jgehrcke** [comment]
@estroz I am genuinely impressed with how nicely you found your way through our chaotic code base :).

Thanks for this porposal.

Sometimes it makes a little more sense to first define 'desired / correct behavior' in a discussion before writing code, and this here might be just one of these cases; because I think we misled you a bit here and I am a little sorry for that.

The fact that we _update_ the controller config as part of `create` seems like this is well thought through and by design; but no. We should not do this; it is a source of problems. Running `create` twice is not generally supported as of today and known to lead to a number of problems, as of a rather huge design gap. Also see https://github.com/opstrace/opstrace/issues/20, a tiny quote:

> this means the same command run twice (unaltered) can't just magically do the right thing in an idempotent fashion. There's a logical conflict here. Even with proper config upgrade mechanism we should not magically switch between 'initial create' and 'config upgrade'.

(note to self: it's really time to make changes to the installer code base to make that clear and to error out)

That is, the idea of setting `lastInstallerVersion` in the controller config as part of `create` is flawed.

But I think we can do a more or less tiny tweak to this PR, and persist the installer version of the _initial_ install:

During initial `create`, i.e. in `installer/index.ts`, instead of `lastInstallerVersion` let's persist the concept of the _initial_ installer version (maybe call this `initialInstallerVersion`). That's useful information.

Then, subsequently, we can add to this structure in the _upgrader_. Not sure if you've seen the upgrader, and how it mutates the controller config:
https://github.com/opstrace/opstrace/blob/2344fa516f6b96b3d7fde54d45b6ccc35e4bba4f/packages/upgrader/src/index.ts#L194

The upgrader itself as of today doesn't really respect much of the input it gets -- most of the user-given instance config -- but at least _by design_ it is supposed to mutate the controller config. It can therefore make sense to introduce a `lastUpgraderVersion`, as part of that mutation.

And honestly, while we are here, instead of doing just `lastUpgraderVersion` we could make each upgrade add to a JSON array (i.e. an ordered structure) and add timestamps; and with that record a history of initial install and all upgrades in the controller version (I can see that this may seem dirty in many ways, but honestly I think this is pragmatic and there will be little practical fallout and it might come in handy every now and then).

**Review by jgehrcke — CHANGES_REQUESTED** [review]
Forgot to leave the previous comment as a review, doing this now :).

**Comment by jgehrcke** [comment]
you motivated me to do this tiny part: https://github.com/opstrace/opstrace/pull/1050

File: packages/controller-config/src/schemav2.ts
Comment (sreis) [inline packages/controller-config/src/schemav2.ts:72]: Because this adds a new field to the v2 schema it'll fail during controller configuration upgrades. The reason being that [upgradeControllerConfigMapToLatest](https://github.com/opstrace/opstrace/blob/e4c1c289ecfb917aca62143d6c942bd7ceffe265/packages/controller-config/src/schema.ts#L79-L107) will not be able to validate the (now old) v2 schema configuration on a running instance because the `installerMetadata` is now a required field.

One option is to make `installerMetadata` not required the other would be to introduce a new schema version (v3) and a `V2toV3` upgrade function that would set the new field.

Makes sense?

**Review by sreis — COMMENTED** [review]
This is a great contribution and overall I like the direction. 

Unfortunately, we don't have CI configured to run jobs when the PR is from external contributors so I will volunteer to clean up any potential failures in the main branch when this is merged.

File: packages/cli/src/info.ts
Comment (estroz) [reply packages/cli/src/info.ts:53]: The kubeconfig returned from [`getKubeConfigForOpstraceClusterOrDie`](https://github.com/opstrace/opstrace/blob/696f18892cf3ec621129c9054489105265233797/packages/cli/src/util.ts#L303) appears to just generate a kubeconfig dynamically using the same [underlying function](https://github.com/opstrace/opstrace/blob/696f18892cf3ec621129c9054489105265233797/packages/cli/src/util.ts#L348) as [`getKubecfgIfk8sClusterExists`](https://github.com/opstrace/opstrace/blob/main/packages/uninstaller/src/index.ts#L91-L105); will they not yield the same result? Perhaps exiting is not the right behavior though, and returning "undefined" for cluster version for an unavailable cluster using the function you suggested is desirable instead. Thoughts?

File: packages/controller-config/src/schemav2.ts
Comment (estroz) [reply packages/controller-config/src/schemav2.ts:72]: I was concerned that adding a new field would break some contract. Typically when I think of changes that require a major version upgrade, I think of breaking changes to an existing contract; in this case, `installerMetadata` is a net-new field that isn't _required_ for any CLI functionality to work; however I could see some project component eventually relying on it's presence for all v2 configs when that isn't guaranteed if it remains optional.

How about schema v3alpha1, in case `installerMetadata`'s schema is changed soon based on new ideas? Is there a stabilization timeline you typically follow?

**Comment by estroz** [comment]
@jgehrcke thanks for the thorough response. Your clarification on `create`'s intended behavior makes sense. A few questions:

>During initial create, i.e. in installer/index.ts, instead of lastInstallerVersion let's persist the concept of the initial installer version (maybe call this initialInstallerVersion). That's useful information.

Is treating the initial and subsequent installer versions different useful? Assuming all upgrader versions are stored in a time-ordered list, could we not treat the first item in that list the initial installer version and do away with a separate field?

>And honestly, while we are here, instead of doing just lastUpgraderVersion we could make each upgrade add to a JSON array (i.e. an ordered structure) and add timestamps; and with that record a history of initial install and all upgrades in the controller config (I can see that this may seem dirty in many ways, but honestly I think this is pragmatic and there will be little practical fallout and it might come in handy every now and then).

I'm fine with adding each upgrader version to an array, as it may help with debugging. I will note that the max ConfigMap size is [1MiB](https://kubernetes.io/docs/concepts/configuration/configmap/#motivation), which is more than reasonable for a typical config file; however, now we're discussing adding a field whose size scales with number of upgrades, which could cause issues as more config fields are added and the array is not garbage collected. Thoughts on this?

>Cosmetically, about cluster_version: if this is supposed to show the underlying k8s version, then let's maybe call this k8s_version :).

:+1: 

>Unfortunately, we don't have CI configured to run jobs when the PR is from external contributors so I will volunteer to clean up any potential failures in the main branch when this is merged.

@sreis I really appreciate it! I'll attempt to make this as failure/bug-free as possible prior to merge.

**Comment by jgehrcke** [comment]
> Assuming all upgrader versions are stored in a time-ordered list, could we not treat the first item in that list the initial installer version and do away with a separate field?

Yes. How would we call the single list?

> Is treating the initial and subsequent installer versions different useful?

Useful maybe yes. But not essential; from an information / storage point of view, I agree to that.

It may be useful for human interpretation I think: I think I like the idea of using expressive variable names here where it's immediately obvious that there is the "initial installer version" (or something equally expressive) keeping record of the version used for `opstrace create`  and then the "upgrader versions" (or something equally expressive), keeping record of the version(s) used for `opstrace upgrade`.

That is, I think this is 'only' about naming and semantics and expressiveness.



Feel free to make a decision here!

> I will note that the max ConfigMap size is 1MiB, which is more than reasonable for a typical config file; however, now we're discussing adding a field whose size scales with number of upgrades, which could cause issues as more config fields are added and the array is not garbage collected. Thoughts on this?

Let's add garbage collection when it's needed. It's likely never needed. We have a lot of time / headroom to figure that out. Back of the envelope calc: when each `upgrade` entry adds 50 bytes then we have space for about 20000 of those which means more than 350 years of upgrades when one does an upgrade every 7 days.

File: packages/controller-config/src/schemav2.ts
Comment (sreis) [reply packages/controller-config/src/schemav2.ts:72]: > How about schema v3alpha1, in case installerMetadata's schema is changed soon based on new ideas? Is there a stabilization timeline you typically follow?

At this point, we don't have a schedule to handle these schema upgrades. AFAIK, there are no plans to touch the controller config map schema in the short term. You can go ahead and introduce `V3` then.

File: packages/controller-config/src/schemav3.ts
Comment (estroz) [inline packages/controller-config/src/schemav3.ts:69]: @jgehrcke I've made this into an array, and changed the name to `allCLIVersions` since the `opstrace` CLI will always be the agent creating or updating the config. If you have a better suggestion I'll use it, since I do not have a strong opinion here.

File: packages/controller-config/src/schemav2.ts
Comment (estroz) [reply packages/controller-config/src/schemav2.ts]: Done.

File: packages/cli/src/index.ts
Comment (estroz) [reply packages/cli/src/index.ts]: Done.

File: packages/controller-config/src/schemav3.ts
Comment (estroz) [inline packages/controller-config/src/schemav3.ts:55]: Since I'm creating a new config schema, should I handle this now?

**Comment by estroz** [comment]
I was not able to successfully run `create` locally (informer creation kept 404'ing) so new changes have not been tested yet. Will try again with a clean environment soon.

**Review by jgehrcke — APPROVED** [review]
:rocket: I am happy with that current state! @sreis will you take another look w.r.t. the schema versions?

**Comment by sreis** [comment]
> We should also trigger a CI run. :)

Yes, but we disabled running builds from forks.

> I was not able to successfully run create locally (informer creation kept 404'ing) so new changes have not been tested yet. Will try again with a clean environment soon.

Let's wait for further feedback from @estroz on his local testing. Depending on that, we can try to enable builds from PRs and kick off a CI build. When that is finished, we'll need to disable builds from PRs again.

File: packages/upgrader/src/index.ts
Comment (estroz) [inline packages/upgrader/src/index.ts:201]: @sreis was this call added for a reason in https://github.com/opstrace/opstrace/pull/869/files#diff-d85ea48a17112e49ec6714d9c57517cc0192ca603adad37fbb65bc893c52db0bR201? It looks extraneous to me but I can't be sure.

File: packages/controller-config/src/schemav3.ts
Comment (estroz) [inline packages/controller-config/src/schemav3.ts:32]: Had to add this Date parsing workaround.

**Comment by estroz** [comment]
Tested this an seems to work, using explicit image tags so the upgrade checks pass:
```console
$ opstrace create aws test-cluster -c config.yaml
$ kubectl get configmap opstrace-controller-config -o json | jq -r '.data."config.json" | fromjson | .cliMetadata'
{
  "allCLIVersions": [
    {
      "version": "3683e58ec-dev",
      "timestamp": "2021-07-24T21:20:00.332Z"
    }
  ]
}
$ opstrace info aws test-cluster
...
cluster info:
  kubernetes_version: v1.18.16-eks-7737de
  controller_version: 2e4ef97a-ci
  installer_version: 3683e58ec-dev
$ opstrace upgrade aws test-cluster -c config.yaml
$ kubectl get configmap opstrace-controller-config -o json | jq -r '.data."config.json" | fromjson | .cliMetadata'
{
  "allCLIVersions": [
    {
      "version": "3683e58ec-dev",
      "timestamp": "2021-07-24T21:20:00.332Z"
    },
    {
      "version": "3683e58ec-dev",
      "timestamp": "2021-07-24T21:20:05.172Z"
    }
  ]
}
$ opstrace info aws test-cluster
...
cluster info:
  kubernetes_version: v1.18.16-eks-7737de
  controller_version: a0fb5b72-ci
  installer_version: 3683e58ec-dev
```

File: packages/upgrader/src/index.ts
Comment (sreis) [reply packages/upgrader/src/index.ts:201]: No, it's a good catch from you to delete one of the calls.

File: packages/controller-config/src/schemav3.ts
Comment (sreis) [reply packages/controller-config/src/schemav3.ts:77]: I'd say no otherwise this will turn out to be an even bigger PR.

File: packages/cli/src/info.ts
Comment (sreis) [reply packages/cli/src/info.ts:60]: If we can't get the kubeconfig to access the cluster, then we might as well fail early. wdyt?

**Comment by sreis** [comment]
Going to try to get Buidkite to run a CI build for this PR now.

**Comment by sreis** [comment]
Got it running but it failed in the build step:

```
[2021-07-26T18:27:24Z] + make build-and-push-controller-image
...
[2021-07-26T18:31:12Z] src/schemav3.ts(18,26): error TS2307: Cannot find module 'date-fns' or its corresponding type declarations.
```

The changes to the schema affect the controller which you haven't caught since you haven't built the docker image. Try running `make tsc` locally and let me know if it catches the error.

**Comment by estroz** [comment]
@sreis surprisingly `make tsc` works fine in my env, and this wasn't an issue when running `make cli` either. I'm guessing `date-fns` is installed globally somewhere in my env. Will add this dependency and push.

File: packages/cli/src/info.ts
Comment (estroz) [reply packages/cli/src/info.ts:60]: Yep agreed! That function is being used above to instantiate `kcfg`.

**Comment by jgehrcke** [comment]
@triclambert @PatrickHeneise why would commits like this end up on this branch?

https://github.com/opstrace/opstrace/pull/1047/commits/c6dbf871f429e5edf650f75cfa8cc3725179394a

**Comment by PatrickHeneise** [comment]
> @triclambert @PatrickHeneise why would commits like this end up on this branch? Do we need to tame that robot?
> 
> [c6dbf87](https://github.com/opstrace/opstrace/commit/c6dbf871f429e5edf650f75cfa8cc3725179394a)

Strange. The Action should run on branches that are not `main` and only if `/docs` are modified, which isn't the case here. I'll review the trigger conditions, thanks for raising the issue.

**Review by jgehrcke — CHANGES_REQUESTED** [review]
TLDR: for encoding timestamps as text, I'd love to fully specify the time string format and to settle on using `js-joda` for the parsing.

---

Instead of `"must be an ISO date string"` let's maybe lock down a specific format :) More specifically, let's do RFC 3339 
- without the fractional second part
- with the `Z` specifier 

Example: `1990-12-31T23:59:59Z` 

`yyyy-MM-ddTHH:mm:ssZ`


One of the lessons I've learned in the past years is that "an ISO date string" can have a gazillion different shapes, assuming we talk ISO 8601. The separator may be a T or a white space, the fraction of a second maybe present or not; it may have this or that precision -- and the time zone may be encoded in many different way, for example via `Y` or `+00:00`.

For datetime operations including serialization and deserialization we've already bought into `js-joda`, a beautiful JavaScript port of a famous lib from the Java ecosystem. Example:

https://github.com/opstrace/opstrace/blob/5d8a612c14fe042ce219a16e70704b511293b7b4/test/test-remote/testutils/index.ts#L31

I think this would do the validation job:

```javascript
> LocalDateTime.parse("1990-12-31T23:59:59Z", DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'"))
LocalDateTime {
  _date: LocalDate { _year: 1990, _month: 12, _day: 31 },
  _time: LocalTime { _hour: 23, _minute: 59, _second: 59, _nano: 0 }
}
```

When passing bad values, the lib throws nice exceptions, such as
```
DateTimeParseException: Text '1990-12-31T23:59:60Z' could not be parsed: Invalid int value for SecondOfMinute: 60: 1990-12-31T23:59:60Z, at index: 0
-------
Caused by: DateTimeException: Invalid int value for SecondOfMinute: 60
```

One could also parse into a `ZonedDateTime` object via `DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssX")` -- the `X` format specifier however allows for many options, see https://js-joda.github.io/js-joda/manual/formatting.html#format-patterns -- `Z; -08; -0830; -08:30; -083015; -08:30:15;`. That's why I would prefer to _require_ Z, and hence the `LocalDateTime.parse()` proposal above.

I want to add that I thought about this before and then thought 'ah, I keep this thinking for myself'. Now that we talk about this new datetime library `date-fns` and how it was missing in CI I changed my mind :).

**Comment by jgehrcke** [comment]
@estroz do you have a checkbox on the right-hand side here "allow edits from maintainers"? If you check this, I am happy to do the js-joda thing, and I think with that we also get another CI run for free :)

https://docs.github.com/assets/images/help/pull_requests/allow-maintainers-to-make-edits-sidebar-checkbox.png

https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

> Additionally, you can modify an existing pull request to let repository maintainers make commits to your branch.

**Comment by estroz** [comment]
@jgehrcke sounds great, thanks! Yes you should be able to push to this branch.

**Comment by jgehrcke** [comment]
OK I rebased on latest `main` (in upstream i.e. this repo) and removed the 'reading time' commits while I was there. Added my commit for said timestamp string validation (and creation).

I also initiated a CI build through our buildkite pipeline by setting `branch` to `refs/pull/1047/head` -- setting `branch` to `estroz:feat/cli-info-cmd` did not work CC @sreis (not sure which method you used before).

**Comment by jgehrcke** [comment]
> I'll review the trigger conditions, thanks for raising the issue

Thank you @PatrickHeneise and also thanks for the quick response!!

**Review by jgehrcke — APPROVED** [review]
:rocket: 

CI went green. Let's land this, and change all the things we'd still like to improve after that.

Thank you so much @estroz for bearing with us here!

**Comment by estroz** [comment]
@jgehrcke @sreis happy to contribute, thanks for assisting!

**Pull request merged**
